src/item.cpp

```diff
@@ -628,7 +629,8 @@ item &item::ammo_set( const itype_id &ammo, int qty )
     // check ammo is valid for the item
     const itype *atype = item_controller->find_template( ammo );
     if( atype->ammo && ammo_types().count( atype->ammo->type ) == 0 &&
-        !magazine_compatible().count( atype->get_id() ) ) {
+        !magazine_compatible().count( atype->get_id() ) && !( magazine_current() ) &&
```

Comment:
`!( magazine_current() )` is only true when `magazine_current()` returns a null pointer. But the condition continues with dereferencing that pointer in the next check. That won't work.
```suggestion
        !magazine_compatible().count( atype->get_id() ) && magazine_current() &&
```

Reply:
Some thing is wrong, I think what I made is
`if( atype->ammo && ammo_types().count( atype->ammo->type ) == 0 &&
        !magazine_compatible().count( atype->get_id() ) && !( magazine_current()  &&
        magazine_current()->ammo_types().count( atype->ammo->type ) ) )`

Need to wash my face and get some fresh air...

Reply:
I wanted to explain again
We want to ckeck if the "ammo" is available.
1. "ammo" is not right ammo for the item.
2. "ammo" is not a magazine for the item.
the 3rd condition I want to make is
the item has no magazine or the  "ammo" is not suitable for the magazine
that is `!( magazine_current() && magazine_current()->ammo_types().count( atype->ammo->type ) )`
which is aquivalent to `!( magazine_current() || magazine_current()->ammo_types().count( atype->ammo->type )`.

both ways, if  magazine_current() is false the next condition won't checked so it's safe

Reply:
You had parentheses put in wrong places then as you had:

`!( magazine_current() ) && magazine_current()->ammo_types().count( atype->ammo->type )`

which would cause crashes as `magazine_current()->ammo_types()` would only be called when `magazine_current()` is `nullptr`.

Reply:
Yes I know, sorry for being incautious for that obvious mistake. I just want to explain that I am not agree with the suggested change and what I actually should have done.